sw/device/lib/runtime/pmp.c

```diff
@@ -0,0 +1,539 @@
+// Copyright lowRISC contributors.
+// Licensed under the Apache License, Version 2.0, see LICENSE for details.
+// SPDX-License-Identifier: Apache-2.0
+
+#include "sw/device/lib/runtime/pmp.h"
+
+#include <stdbool.h>
+#include <stddef.h>
+#include <stdint.h>
+
+#include "sw/device/lib/base/bitfield.h"
+#include "sw/device/lib/base/stdasm.h"
+
+// "Volume II: RISC-V Privileged Architectures V20190608-Priv-MSU-Ratified",
+// "3.6.1 Physical Memory Protection CSRs",
+// "Figure 3.28: PMP configuration register format".
+#define PMP_CFG_CSR_R 0
+#define PMP_CFG_CSR_W 1
+#define PMP_CFG_CSR_X 2
+#define PMP_CFG_CSR_A 3
+#define PMP_CFG_CSR_L 7
+
+#define PMP_CFG_FIELDS_PER_REG 4
+#define PMP_CFG_FIELD_WIDTH 8
+#define PMP_CFG_FIELD_MASK 0xffff
+
+// "Volume II: RISC-V Privileged Architectures V20190608-Priv-MSU-Ratified",
+// "3.6 Physical Memory Protection", "Address Matching".
+#define PMP_CFG_CSR_MODE_OFF 0
+#define PMP_CFG_CSR_MODE_TOR 1
+#define PMP_CFG_CSR_MODE_NA4 2
+#define PMP_CFG_CSR_MODE_NAPOT 3
+#define PMP_CFG_CSR_MODE_MASK 0x3
+
+// TODO - explanation to why macro magic is necessary (csr must be const expr).
+#define PMP_CSR_WRITE(csr, var) asm volatile("csrw " #csr ", %0;" : : "r"(var))
+
+#define PMP_CSR_READ(csr, var) asm volatile("csrr %0, " #csr ";" : "=r"(var) :)
+
+#define PMP_CSR_RW(access, csr, var)       \
+  do {                                     \
+    if (access == kPmpCsrAccessTypeRead) { \
+      PMP_CSR_READ(csr, var);              \
+    } else {                               \
+      PMP_CSR_WRITE(csr, var);             \
+    }                                      \
+  } while (false)
+
+typedef enum pmp_csr_access_type {
+  kPmpCsrAccessTypeRead = 0,
+  kPmpCsrAccessTypeWrite,
+} pmp_csr_access_type_t;
+
+static const bitfield_field32_t kPmpCfgModeField = {
+    .mask = PMP_CFG_CSR_MODE_MASK, .index = PMP_CFG_CSR_A,
+};
+
+// TODO - X macros as suggested by @lenary.
+
+// TODO - description.
+static pmp_region_configure_result_t pmp_cfg_csr_rw(
```

Comment:
It would then require 2x functions with a switch statement. @lenary has a pretty good solution with X-macros, which I intend to fix-up in the fallow-up.

Reply:
Sounds good, we can continue this conversation once this starts using x-macros.